### ISSUE TITLE
Implement more types of tags to enable better photo organization

### ISSUE THREAD

**Issue by jerrac** [darktable-org/darktable#12473]
**Is your feature request related to a problem? Please describe.**
I've been trying to find a good photo management tool for my mom. She's currently using LR 5.7, and it's showing it's age. Long story short, I've been through 10+ different tools, paid and free, and none of them would actually be better for her than LR... They all suffer from poor UI. I'm including darktable and LR in that. As well as ACDsee, PhotoPrism, Microsoft Photo, ON1, Luminar, Capture One, etc.

Ultimately, I've come to the conclusion that I'd like to just have her use darktable. It's UI has issues, but I can see her using it if she puts some effort in. 

But there is one missing feature that she would not be able to do without, an "Album" feature. And that's a feature I would very much appreciate having as well.

I did look for existing issues and found these:

* https://github.com/darktable-org/darktable/issues/10643
* https://github.com/darktable-org/darktable/pull/5134
* https://github.com/darktable-org/darktable/pull/10733

I don't think this post really fits as a comment on any of those issues. Apologies if I'm wrong on that.

**Describe the solution you'd like**
To start, I'd like to talk about how I think about "collections", "categories", "albums" and "tags/keywords." Actually defining what we mean by these different terms might help explain why darktable needs some changes related to them.

* "tags/keywords" are short descriptions of the photo's content. "tree", "dark", "animal", etc.

* "categories" are a broad classifications of many photos. So, "family photos", "wedding photos", "studio photos", etc. would be categories. Yes, those could also be tags, but they have a much broader meaning than what a tag would contain. They describe groups, not items.

* "albums" are a group of photos meant for a specific purpose. So, a scrapbook, a photo book, a set of photos for a client, etc. It's not a tag because it is describing many photo's. It isn't a category because it has a specific purpose/focus.

* "collections" (as a concept) are completely arbitrary groups of photos. You manually add photos to a collection based on any criteria you want. Basically, if it's not a "tag" or an "album" it's a collection.

* "automatic collections" are rules based collections. So, if you need to quickly find a set of photos that were taken by a specific lens, are tagged with "super", and are not used in any "album", you'd make an automatic collection that takes those criteria and will automatically add any photo that meets them to itself.

So, in my opinion, all five items are significantly different enough that they should have their own UI elements. But they are all similar enough that I think we could build on the current tags based structure to get what we need.

How could we implement these concepts? 

* Make "album", and "collection" tag types like we already have a "category" tag type.
* Add "category tags", "album tags", and "collection tags", options to the "collections" dropdown menu.
* Add "set as category", "set as album", and "set as collection", to the tag right click menu.
* I'm not sure on this one, but would it make sense to start calling normal tags "keyword tags"? Would that help keep the use of normal tags clear when we have all the other kinds of tags?
* "automatic collections" are already pretty much implemented by collection filters. But if you end up with a LOT of them, the presets menu is going to get really clunky. So I'd suggest adding a list element to the widget with all the presets in it.

I think that would all fit in with darktable's current UI paradigm.

Personally, I'd take it a bit farther and add widgets for each method of organization. The collections widget is just much too dense. I have to concentrate to really pick the option I'm after. And there are people who struggle with that even more than I do. For example, my Dad has issues with too much information on a screen, it really makes it hard for them to read or see things. He has to make sure to only keep one window open at a time. He'll actually minimize his current window before opening another. So for people with similar issues, the density of the current UI is problematic. 

I'd like to at least split the collections widget up based on files/metadata/etc. categories in the dropdown. Even better would be to use the same kind of structure as the darkroom right hand sidebar with the overall categories of different modules. Then we could have a widget per collections dropdown item.

**Alternatives**
Use the "category" tags as is. Unfortunately, I can't see Mom being able to keep the difference between a category tag and a normal tag separate. There is no significant difference between the two types in the UI. And creating category tags is hidden in the tag edit modal. Mom would forget how to find that very quickly. Heck, I forgot about it until I went looking for Album issues today... 

**Additional context**
Thanks in advance for any consideration. I'm hopeful my ideas are worthwhile, and might at least provoke some constructive discussion.

I don't know C, so I can't really try to help with anything needing that kind of development (I'm a Linux sysadmin/web dev, not much experience with compiled languages), but I do plan to revisit Lua and see if any of my ideas can be implemented that way. 

Also, I just have to say thanks to all the devs who have built darktable! I find it to be the best tool for photo developing and management out there. So I'm definitely grateful. :)

**Comment by ptilopteri** [comment]
know that one can export a photo/group of photos to an "album" by
selecting "webside gallery" in the export module.

if that is satisfactory, please consider closing this issue.

**Comment by ralfbrown** [comment]
@ptilopteri, OP is asking for a way to have a collection spanning multiple directories with arbitrary photos belonging to that collection.

The current dt way is to have a tag for that, applied to the photos one wants to have in the album.  I would suggest using `Album` or `Albums` as the top level of a (probably single-level) tree of tags meant to specify membership in particular albums.  Then selecting the appropriate album tag in the Collections module will show the contents of the album.

All collections in dt are rule-based ('smart') and update immediately on changing any of the selection criteria on any image.  So if you are displaying an album by using an appropriate album tag, removing that tag from any of the displayed images makes that image disappear from the collection immediately.

AFAIK there is no difference between "normal" and category tags when used for selecting the contents of a collection.

**Comment by jerrac** [comment]
@ptilopteri Er, exporting is not what I am after. Organizing within darktable is.

@ralfbrown Yes, the current way is a tag. If you don't want the tag exported, you use category tags. And I do use that method. What I'm asking for is a better UI for it, and to expand that to cover the different kinds of "collections" I defined.

**Comment by xarthurx** [comment]
I found this issue based on similar needs, is there any solution to it?

**Comment by Nilvus** [comment]
> I found this issue based on similar needs, is there any solution to it?

As all open-source software, a solution will come IF someone has skills to do and WANT spent time on it. Before that, there's some possibilities to deal with actual features. There's an album possibility in tag module and possibilities to use tags to create sort-of albums and filter by tag filter. It's not exactly the same but near. I will not explain it here as this place is not a help forum. Many infos are on the manual.

**Comment by jerrac** [comment]
So, this came up in my github notifications and I started thinking about it... Two questions:

A) Is it even possible to implement my ideas via Lua plugins?

B) What if I wrote a companion app in something other than Lua that could interact with darktable via XMP and/or the sqlite3 db's? (I have yet to find time to learn enough Lua to try it that way...) What pitfalls might I encounter?

**Comment by wpferguson** [comment]
> A) Is it even possible to implement my ideas via Lua plugins?

Yes :smiley:

You could use tags as the mechanism for organizing into albums, then put an "albums" module to interact (create, read, update, destroy).

> B) What if I wrote a companion app in something other than Lua that could interact with darktable via XMP and/or the sqlite3 db's? 

You could modify the XMP files outside of darktable. You would have to configure darktable to check for updated XMP files on start and import them.

**Comment by jerrac** [comment]
er, so, I'm 90% certain I can't do what I want in lua. I do not see any indication that I can add a new type of tag in the api docs. [here](https://darktable-org.github.io/luadocs/lua.api.manual/darktable/darktable.tags/) or [here](https://darktable-org.github.io/luadocs/lua.api.manual/types/dt_lua_tag_t/).

If I had to guess, adding a new type of tag would actually involve adding something to [this](https://github.com/darktable-org/darktable/blob/4f3d272415854cc9660af78bba7e1ee84b7893b4/src/common/tags.h#L36) enum, and then making sure all the code that uses it was modified to show the new option.

(By new type, I mean add an option the tag edit dialog like the existing "private" and "category" checkboxes.)

Unless you were thinking I'd be using hierarchical tags and then writing widgets for them? I'll have to think about that approach...

**Comment by wpferguson** [comment]
You can create UI elements using the Lua API (modules, widgets, etc.).  You have almost unlimited control of the UI from Lua.  

Build an Album UI using Lua and tags, starting with a top level tag of "Album".  Take that as far as you can until you run into something it can't do with the existing tag structure.  At that point submit an issue saying what you want to do and why the current tagging system can't satisfy it and ask for help.

**Comment by jerrac** [comment]
I was unable to make enough progress on this to even have a clue what the api might, or might not, be missing. 

As far as I can tell, you cannot create a tagging widget via the api. But I'm not 100% sure on that. None of the existing scripts did anything like a tagging widget that I could see.

I did figure out how to set and retrieve preferences and was experimenting with different widgets. 

At this point I'm a bit frustrated. Reading the descriptions of each widget, and trying to use them, hasn't been enough to help me put the pieces together into what I want to build.

Nor has reading through the existing scripts. I at least skimmed all of them.

So, is creating a tagging widget even possible via Lua?

Has anyone written and up to date tutorial? Something more than just the example scripts. Like, say, "how to make a custom website gallery."

Has anyone written up an example script, or blog post, going over each widget type, what they are, and how they should be used? 

While I would appreciate answers to my questions, I'm also just going to go ahead and close this issue. There doesn't seem to be enough interest in my idea for anyone else to implement it, and I just don't have the time to dig into Lua and darktable enough to do it myself, when I can't get past the issues I already described.
